src/core/algorithms/algorithm.cpp

```diff
@@ -90,13 +92,13 @@ void Algorithm::SetOption(std::string_view option_name, boost::any const& value)
     auto it = possible_options_.find(option_name);
     if (it == possible_options_.end()) {
         if (ext_opt_set) return;
-        throw std::invalid_argument("Unknown option \"" + std::string{option_name} + '"');
+        throw config::UnknownOption("Unknown option \"" + std::string{option_name} + '"');
     }
     std::string_view name = it->first;
     config::IOption& option = *it->second;
     if (available_options_.find(name) == available_options_.end()) {
         if (ext_opt_set) return;
-        throw std::invalid_argument("Invalid option \"" + std::string{name} + '"');
+        throw config::UnknownOption("Invalid option \"" + std::string{name} + '"');
```

Comment:
There is the same exception type but different error messages because the errors are slightly logically different. Maybe we should create a separate exception type for existing but not available options?

Reply:
Now that I think of it, is there any situation where one would need to catch a particular type of configuration exception? Perhaps there should only be one `ConfigurationError`. What are your thoughts?

Reply:
Well, I'm not sure about a possible usecase where we want to catch only a specific type of configuration related exceptions, but this doesn't seem unrealistic.. Also, separate exception types reduce amount of boilerplate code which generates exception messages.
However, all this is rather minor and right now there is no need to handle different configuration errors differently. I think it's acceptable to use only `ConfigurationError` and get rid of all other now redundant types. Other exceptions can be added when they're needed.